lib/puppet/util/windows/eventlog.rb

```diff
@@ -0,0 +1,180 @@
+require 'ffi'
+
+# Puppet::Util::Windows::EventLog needs to be requirable without having loaded
+# any other parts of Puppet so it can be leveraged independently by the code
+# that runs Puppet as a service on Windows.
+#
+# For this reason we:
+# - Define Puppet::Util::Windows
+# - Replicate logic that exists elsewhere in puppet/util/windows
+# - Raise generic RuntimeError instead of Puppet::Util::Windows::Error if its not defined
+module Puppet; module Util; module Windows ; end ; end ; end
+
+class Puppet::Util::Windows::EventLog
+  extend FFI::Library
+
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa363679(v=vs.85).aspx
+  EVENTLOG_ERROR_TYPE       = 0x0001
+  EVENTLOG_WARNING_TYPE     = 0x0002
+  EVENTLOG_INFORMATION_TYPE = 0x0004
+
+  # Register an event log handle for the application
+  # @param source_name [String] the name of the event source to retrieve a handle for
+  # @return [void]
+  # @api public
+  def initialize(source_name = 'Puppet')
+    @eventlog_handle = RegisterEventSourceW(nil, wide_string(source_name))
+    if @eventlog_handle == NULL_HANDLE
+      raise EventlogError.new("failed to open Windows eventlog")
+    end
+  end
+
+  # Close this instance's event log handle
+  # @return [void]
+  # @api public
+  def close
+    result_of_close = DeregisterEventSource(@eventlog_handle)
+    if result_of_close == WIN32_FALSE
+      raise EventlogError.new("failed to close Windows eventlog")
```

Comment:
I don't see anywhere in the code that we actually call close?  So I have a few thoughts on how you might handle this:

* Register some sort of `at_exit` handler like we do in TaskScheduler - https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/windows/taskscheduler.rb#L194-L200 - noting the problems with `at_exit`.  For one, you can't guarantee ordering of `at_exit` handlers IIRC, so it's plausible that something could still try to use the event log after the handle has been closed. If the code was made more defensive, we could prevent that from being a problem *BUT* we might mysteriously lose event log messages, which could be tricky to diagnose.
* Don't store the handle in `@eventlog_handle` at all - instead grab a new one before each logging attempt, and  then close it immediately after use.  There will be some additional overhead, but given our limited usage of `report_event`, this doesn't seem so terrible to me.
* Don't worry about ever calling `close` - when the process closes it will eventually close down any open kernel objects (including handles), and I don't think there will be any real-world negative consequences.